crates/accelerate/src/commutation_cancellation.rs

```diff
@@ -0,0 +1,291 @@
+                    for &c_node in &cancel_set[0..(cancel_set.len() / 2) * 2] {
+                        dag.remove_op_node(c_node);
+                    }
+                }
+            }
+            if matches!(cancel_key.gate, GateOrRotation::ZRotation) && z_var_gate.is_none() {
+                continue;
+            }
+            if matches!(cancel_key.gate, GateOrRotation::ZRotation)
+                || matches!(cancel_key.gate, GateOrRotation::XRotation)
+            {
+                let run_op = match &dag.dag[*cancel_set.first().unwrap()] {
+                    NodeType::Operation(instr) => instr,
+                    _ => panic!("Unexpected type in commutation set run."),
+                };
+
+                let run_qarg = dag.get_qargs(run_op.qubits).first().unwrap();
+                let mut total_angle: f64 = 0.0;
+                let mut total_phase: f64 = 0.0;
+                for current_node in cancel_set {
+                    let node_op = match &dag.dag[*current_node] {
+                        NodeType::Operation(instr) => instr,
+                        _ => panic!("Unexpected type in commutation set run."),
+                    };
+                    let node_op_name = node_op.op.name();
+
+                    let node_qargs = dag.get_qargs(node_op.qubits);
+                    if node_op
+                        .extra_attrs
+                        .as_deref()
+                        .is_some_and(|attr| attr.condition.is_some())
+                        || node_qargs.len() > 1
+                        || &node_qargs[0] != run_qarg
+                    {
+                        panic!("internal error");
+                    }
+
+                    let node_angle = if ROTATION_GATES.contains(&node_op_name) {
+                        match node_op.params_view().first() {
+                            Some(Param::Float(f)) => *f,
+                            _ => return Err(QiskitError::new_err(format!(
+                                "Rotational gate with parameter expression encoutned in cancellation {:?}",
+                                node_op.op
+                            )))
+                        }
+                    } else if HALF_TURNS.contains(&node_op_name) {
+                        PI
+                    } else if QUARTER_TURNS.contains(&node_op_name) {
+                        PI / 2.0
+                    } else if EIGHTH_TURNS.contains(&node_op_name) {
+                        PI / 4.0
+                    } else {
+                        panic!("Angle for operation {node_op_name} is not defined")
+                    };
+                    total_angle += node_angle;
+
+                    if let Some(definition) = node_op.op.definition(node_op.params_view()) {
+                        total_phase += match definition.global_phase() {Param::Float(f) => f, _ => panic!("PackedInstruction with definition has no global phase set as floating point number")};
+                    }
```

Comment:
As the logic is right now this should return a python error:

```suggestion
                        total_phase += match definition.global_phase() {
                            Param::Float(f) => f,
                             _ => return Err(QiskitError::new_err(format!("PackedInstruction with definition has no global phase set as floating point number")))
                        };
                    }
```

Although realistically if you make my above suggestion about how to get the definition directly from `StandardGate` I think you can change this to just use `unreachable!()` since the generator code for definitions of standard gates never has a parameterized phase, it's always a float `0.0`. Something like:

```suggestion
                        let Param::Float(new_phase) = definition.global_phase() else { unreachable!() };
                        total_phase += new_phase
```
(although maybe set an error message saying standard gate phase is never parameterized)

would do the trick. But we can't make this change unless we're working with a `StandardGate` when we call `definition()`.

Reply:
Unless I'm missing something: one of Rz or P (whichever isn't exactly equal to u1) can have a parametric phase if the angle is parametric. `GlobalPhaseGate` definitely can (if the angle is parametric).

Reply:
Ah good point, I thought there was one, but I scanned through the match statement yesterday and missed it on rzgate. But regardless this statement still holds since the angle is always a float at this point